### DISCUSSION TITLE
Iterable jobs: pass iteration arguments to `around_iteration` as well

### DISCUSSION THREAD

**Discussion by gstokkink-bex** [sidekiq/sidekiq#6453] [Ideas]
We want to add some instrumentation to our iterable jobs that tracks some metrics for each iteration. It would be extremely useful for us if we could also gain access to the iteration arguments (`object` and `arguments`). Right now `iterate_with_enumerator` directly calls `around_iteration` without passing the arguments: https://github.com/sidekiq/sidekiq/blob/main/lib/sidekiq/job/iterable.rb#L162.

**Comment by mperham** [comment]
I've found that I dislike the current design which makes the user save any arguments manually.

What do you think about an `arguments` reader which allows any callback to access the job's (frozen) arguments, including `around_iteration`?

**Reply by gstokkink-bex** [reply]
@mperham that sounds good, but it would be nice to be able to access `object` in `around_iteration` as well. That one would have to be passed, I think.

Right now we work around the lack of visibility of the job arguments (and other relevant attributes) inside of `around_iteration` by defining a `job_payload` (i.e., `msg`) accessor and setting `job_instance.job_payload = job_payload` in a server middleware. Works well but feels hacky.

**Reply by mperham** [reply]
I'm not sure. `each_iteration` is called right after `around_iteration` with the object so the two methods start to look identical/redundant. Give me a good usecase?

**Reply by mperham** [reply]
See #6456

**Reply by gstokkink-bex** [reply]
@mperham amongst other things, we want to be able to report soft timeout notifications when an iteration's runtime exceeds 25 seconds, so the developers can fix that. To be able to fix such a slow iteration, it's immensely helpful to be able to pinpoint which iteration exactly was too slow. For that, you need the `object` I think, as there is no 'index' being kept track of that can be reported.

**Reply by mperham** [reply]
I think that's a use case which should be built into Sidekiq, similar to the Redis RTT warning in sidekiq/launcher.rb.

**Reply by gstokkink-bex** [reply]
@mperham if you're going for the built-in approach it would be nice if the measurements/warnings could be instrumented. For us it would not be very useful if the measurements/warnings would simply be written to a logger (like the RTT warnings). Instead, we'd like a notification to be posted using `Rails.error.report`. Instrumentation support would allow anyone to hook up their own logging/notification facilities.

By the way, we also run a very simple custom server middleware that measures (non-iterable) job runtimes and sends a notification if the runtime exceeds 25 seconds. Perhaps also useful as a built-in functionality?
